dbms/src/Common/RWLock.cpp

```diff
@@ -201,33 +203,33 @@ RWLockImpl::LockHolderImpl::~LockHolderImpl()
 {
     std::lock_guard lock(parent->mutex);
 
-    /// Remove weak_ptrs to the holder, since there are no owners of the current lock
-    parent->query_id_to_holder.erase(query_id);
-
-    if (*it_client == RWLockImpl::Read && query_id != RWLockImpl::NO_QUERY)
-        all_read_locks.remove(query_id);
+    if (query_id != RWLockImpl::NO_QUERY)
+    {
+        if (--parent->owner_queries.at(query_id) == 0)
+            parent->owner_queries.erase(query_id);
 
-    /// Removes myself from client list of our group
-    it_group->clients.erase(it_client);
+        if (it_group->type == RWLockImpl::Read)
+            all_read_locks.remove(query_id);
+    }
 
-    /// Remove the group if we were the last client and notify the next group
-    if (it_group->clients.empty())
+    /// Remove the group if we are the last referer and notify the next group
```

Comment:
Contrary to the HTTP standard, the `referrer` word is written with two `r`.

Reply:
Good point. Corrected this